File: Programmers/cpp/LEVEL_2_택배상자.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(vector<int> order) {
	int answer = 0;
	int idx = 0;
	stack<int> st;

	for (int i = 1; i <= order.size(); i++) {
		if (i != order[idx]) {
			if (st.empty() || st.top() != order[idx]) {
				st.push(i);
				continue;
			}

			while (!st.empty() && st.top() == order[idx]) {
				answer++;
				idx++;
				st.pop();
			}
			continue;
		}

		answer++;
		idx++;

		while (!st.empty() && st.top() == order[idx]) {
			answer++;
			idx++;
			st.pop();
		}
	}

	return answer;
}

File: Programmers/java/LEVEL_2_택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int idx = 0;
        Stack<Integer> stack = new Stack<>();

        for (int i = 1; i <= order.length; i++) {
            if (i != order[idx]) {
                if (stack.isEmpty() || stack.peek() != order[idx]) {
                    stack.push(i);
                    continue;
                }

                while (!stack.isEmpty() && stack.peek() == order[idx]) {
                    answer++;
                    idx++;
                    stack.pop();
                }
                continue;
            }

            answer++;
            idx++;

            while (!stack.isEmpty() && stack.peek() == order[idx]) {
                answer++;
                idx++;
                stack.pop();
            }
        }

        return answer;
    }
}